internal/tfdiags/diagnostic_base.go

```diff
@@ -38,3 +38,23 @@ func (d diagnosticBase) FromExpr() *FromExpr {
 func (d diagnosticBase) ExtraInfo() interface{} {
 	return nil
 }
+
+func (d diagnosticBase) Equals(otherDiag ComparableDiagnostic) bool {
+	od, ok := otherDiag.(diagnosticBase)
+	if !ok {
+		return false
+	}
+	if d.severity != od.severity {
+		return false
+	}
+	if d.summary != od.summary {
+		return false
+	}
+	if d.detail != od.detail {
+		return false
+	}
+	if d.address != od.address {
```

Comment:
Address was added to give some extra context to an error to help the user easily identify the resource, but wasn't intended to guarantee uniqueness, isn't required, and may or may not contain multiple levels of expansion depending when it was generated. We should probably be comparing source location rather than the string format here to correlate when errors are coming from the exact same location in the configuration. This would also be how we need to compare hcl diagnostics so we're not coalescing diagnostics from different parts of the configuration.

Reply:
I agree - I will add the source range to comparison function.

Note that `diagnosticBase` specifically does not have the source range though and we still use it here https://github.com/hashicorp/terraform/blob/8f9ef6b8e8d3f2f9a45f723f54ed8b61846f0a32/internal/tfdiags/sourceless.go#L10-L16

I assume that's fine though because we have nothing else to compare anyway.

Reply:
I added source range to the comparisons now but left the address in as it won't make much difference performance wise and two of the same diagnostics should have matching address anyway.

Reply:
The address will actually prevent you from detecting duplicates, but any expanded resource will have at least 2 different address representations for the same diagnostics between initial validation, plan and apply.

Reply:
OK, I understand the full implications now. Thanks.

I have replaced the comparisons with a comment.